### PR TITLE
implementation (Languages): #28477 Add / Edit new Language 

### LINKED ISSUE #28477: Add / Edit new Language
### Parent Issue

#28474

### User Story

As a user, I need the ability to add new Locale for eventual listing. This will provide me with a comprehensive view of the available languages within the system and empower me to contribute new languages as needed.

### Acceptance Criteria

- [ ] Users should be able to add new locales via a user-friendly interface integrated into the listing table.
- [ ] Follow Designs. 

### Proposed Objective

User Experience

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

https://www.figma.com/file/tTWRVYunjnzTAWaj8Ctvra/Locales-%2F-Languages?node-id=472-21986&mode=dev

DEMO: https://www.figma.com/proto/tTWRVYunjnzTAWaj8Ctvra/Locales-%2F-Languages?page-id=0%3A1&type=design&node-id=356-41347&viewport=-1747%2C-4187%2C0.26&t=fx8jP5j5bAGXSHbC-1&scaling=scale-down&starting-point-node-id=356%3A41347&show-proto-sidebar=1

Reference: https://github.com/dotCMS/core/issues/25605 

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

### PR DESCRIPTION
**Pull request by hmoreras**
### Proposed Changes
* Add a Locale. 
* Add a Custom Local 
* Edit a Locale

### Screenshots

https://github.com/dotCMS/core/assets/31667212/f57677d9-9cab-4a2b-a413-4b73d9f10bd2

### CODE REVIEW

File: core-web/libs/dotcms-scss/angular/_forms.scss [hunk 1]
```diff
@@ -60,6 +60,7 @@
     display: block;
     font-size: $font-size-sm;
     margin-top: 0.25rem;
```
Comment (zJaaal) [inline core-web/libs/dotcms-scss/angular/_forms.scss:62]: I think we have a variable for this value

File: core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+            expect(spectator.component.data).toEqual({
+                languages: mockLanguagesISO.languages,
+                countries: mockLanguagesISO.countries,
+                locale: mockLocales[0],
+                localeList: mockLocales
+            });
+        });
+
+        it('should load the edit form correctly', () => {
+            //  console.log('langEdit: ', spectator.query(byTestId('langEdit')));
```
Comment (zJaaal) [inline core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.spec.ts:67]: delete

File: core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    handleSubmit(): void {
+        if (this.form.valid) {
+            const { language, languageCode, country, countryCode } = this.form.value;
+            const isDuplicate = this.data.localeList.some(
+                (locale) => locale.isoCode?.toLowerCase() === this.getISOCode()
+            );
+
+            if (this.data.locale?.id || (!this.data.locale?.id && !isDuplicate)) {
```
Comment (zJaaal) [inline core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts:79]: I think this is the same as `this.data.locale?.id || !isDuplicate`

**Review by zJaaal — CHANGES_REQUESTED** [review]

File: core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    ];
+
+    ngOnInit(): void {
+        this.initForm();
+    }
+
+    handleSubmit(): void {
+        if (this.form.valid) {
+            const { language, languageCode, country, countryCode } = this.form.value;
+            const isDuplicate = this.data.localeList.some(
```
Comment (oidacra) [inline core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts:75]: What about 
```
private isDuplicateLocale(localeList: any[], isoCode: string): boolean {
    return localeList.some((locale) => locale.isoCode?.toLowerCase() === isoCode.toLowerCase());
}
```
and

```
private shouldCloseDialog(localeId: string | null, isDuplicate: boolean): boolean {
    return !!localeId || (!localeId && !isDuplicate);
}
```
More clearly what are you doing there The Beto of the future will thank you for that.

File: core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            countryCode: new FormControl(this.data.locale?.countryCode),
+            id: new FormControl({ value: this.data.locale?.id, disabled: true }),
+            isoCode: new FormControl({ value: this.data.locale?.isoCode, disabled: true }),
+            languageDropdown: new FormControl(''),
+            countryDropdown: new FormControl('')
+        });
+
+        this.languageId = this.form.get('id');
+
+        this.form.get('languageDropdown')?.valueChanges.subscribe((language: DotISOItem) => {
```
Comment (oidacra) [inline core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts:116]: Move this to a private method with a self-explanatory name like `private subscribeToLanguageDropdownChanges()` and the same for `private subscribeToCountryDropdownChanges()` and remember both unsubscribe.

File: core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+import { DotMessagePipe } from '@dotcms/ui';
+
+export interface DotLocaleCreateEditData {
+    languages: DotISOItem[];
+    countries: DotISOItem[];
+    locale: DotLanguage | null;
+    localeList: DotLanguage[];
+}
+
+@Component({
```
Comment (oidacra) [inline core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts:32]: Add JSDoc, to explain what this component is.

File: core-web/libs/dotcms-models/src/lib/dot-language.model.ts [hunk 7]
```diff
@@ -1,6 +1,6 @@
 export interface DotAddLanguage {
-    country: string;
-    countryCode: string;
+    country?: string;
```
Comment (oidacra) [inline core-web/libs/dotcms-models/src/lib/dot-language.model.ts:2]: Why optionals?

Comment (hmoreras) [reply core-web/libs/dotcms-models/src/lib/dot-language.model.ts:2]: I order to create a new one, only the language and languageCode are required. others are optional.

File: core-web/libs/dotcms-scss/angular/_forms.scss (hunk 1 above)
Comment (hmoreras) [reply core-web/libs/dotcms-scss/angular/_forms.scss:62]: changed !

File: core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts (hunk 3 above)
Comment (hmoreras) [reply core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts:79]: not really.. moving those conditions to functions to be more descriptive.

File: core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts (hunk 4 above)
Comment (hmoreras) [reply core-web/libs/portlets/dot-locales/portlet/src/lib/dot-locales-list/components/dot-locale-create-edit/DotLocaleCreateEdit.component.ts:75]: agree ! :)

**Review by oidacra — APPROVED** [review]

**Review by zJaaal — APPROVED** [review]

File: core-web/libs/dotcms-models/src/lib/dot-language.model.ts [hunk 8]
```diff
@@ -1,6 +1,6 @@
 export interface DotAddLanguage {
```
Comment (jdotcms) [inline core-web/libs/dotcms-models/src/lib/dot-language.model.ts:1]: isnt called locale?

File: core-web/libs/data-access/src/lib/dot-languages/dot-languages.service.ts [hunk 9]
```diff
@@ -65,15 +65,27 @@ export class DotLanguagesService {
...
      * Update a language.
      *
-     * @param {DotLanguage} language - The language to be updated.
+     * @param {DotLanguage} locale - The language to be updated.
      * @return {Observable<DotLanguage>} An observable of the updated language.
      */
-    update(language: DotLanguage): Observable<DotLanguage> {
+    update(locale: DotLanguage): Observable<DotLanguage> {
+        const { id, languageCode, language, countryCode, country } = locale;
+
```
Comment (jdotcms) [inline core-web/libs/data-access/src/lib/dot-languages/dot-languages.service.ts:86]: here may need some validations?

File: core-web/libs/dotcms-models/src/lib/dot-language.model.ts (hunk 8 above)
Comment (hmoreras) [reply core-web/libs/dotcms-models/src/lib/dot-language.model.ts:1]: right, the interface kept as it was, like the Database table, the `Locale` is visual.

**Pull request merged**